File: lib/02.dart
int A02(String contents) {
  RegExp parser = RegExp(r'(\d+)\-(\d+)\s([a-z])\:\s([a-z]+)');

  Iterable<String> list = contents.split('\n').where((val) {
    int min = int.parse(parser.firstMatch(val).group(1));
    int max = int.parse(parser.firstMatch(val).group(2));
    String char = parser.firstMatch(val).group(3);
    String str = parser.firstMatch(val).group(4);
    int len = str.split('').where((ch) { return ch == char; }).length;
    return len >= min && len <= max;
  });

  return list.length;
}

int B02(String contents) {
  RegExp parser = RegExp(r'(\d+)\-(\d+)\s([a-z])\:\s([a-z]+)');

  Iterable<String> list = contents.split('\n').where((val) {
    int a = int.parse(parser.firstMatch(val).group(1));
    int b = int.parse(parser.firstMatch(val).group(2));
    String char = parser.firstMatch(val).group(3);
    String str = parser.firstMatch(val).group(4);
    return (str.substring(a - 1, a) == char && str.substring(b - 1, b) != char) || (str.substring(a - 1, a) != char && str.substring(b - 1, b) == char);
  });

  return list.length;
}

File: lib/02.js
module.exports = {
	A02(contents) {
		const lines = contents
			.split(/\n/g)
			.map((line) => {
				return line.match(/(\d+)\-(\d+)\s([a-z])\:\s([a-z]+)/).slice(1, 5);
			})
			.filter(([min, max, char, str]) => {
				const length = str.split('').filter((ch) => ch === char).length;
				return length >= min && length <= max;
			});

		return lines.length;
	},

	B02(contents) {
		const lines = contents
			.split(/\n/g)
			.map((line) => {
				return line.match(/(\d+)\-(\d+)\s([a-z])\:\s([a-z]+)/).slice(1, 5);
			})
			.filter(([a, b, char, str]) => {
				return (str[a - 1] === char) ^ (str[b - 1] === char);
			});

		return lines.length;
	},
};
